### PR TITLE
[webview_flutter] add events: onPageStarted, onReceivedError (with some simple error handling) and action: stopLoading.

### PR DESCRIPTION
**Pull request by hpoul**
## Description

* Adds new events to WebView for `onPageStarted` and `onReceivedError`
   * Tries to combine error handling between iOS/Android
* add `stopLoading` of webview.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by amirh** [comment]
@hpoul please remove WIP label when the PR is ready for review.

**Comment by hpoul** [comment]
@amirh there are just two things I am unsure before it's ready imo:

1. unifying android/iOS - WKWebView and android's WebView present errors quite differently (and the iOS documentation is pretty sparse about which error callback is called when) - I have tried my best to unify the logic. The question is if that is preferable to just passing iOS and Android specifics back to flutter (like e.g. device_info would pass an iosInfo and androidInfo depending on platform.. I hope I chose a useful subset of errors which can occur, but there might be edge cases where it could be useful to have specific errors.. (although I guess it should be easy to add more error cases later in a backward compatible way)
2. tests: I have added a simple test for `onPageStarted` .. maybe you could give me a hint on how to add e2e tests? will take a look tomorrow if there are any examples..

**Comment by hpoul** [comment]
@amirh maybe you could give a short comment if this implementation could fit into the project, then i could start working on basing basic http auth support based on this PR.. (because supporting authentication requires error handling callbacks and I wouldn't want to completely refactor it again..) thanks. :)

(note: ups, had that comment posted in the wrong PR previously, sorry about that)

File: packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebView.java [hunk 1]
```diff
@@ -128,6 +131,12 @@ private void goForward(MethodCall methodCall, Result result) {
   private void reload(MethodCall methodCall, Result result) {
     webView.reload();
     result.success(null);
+    webView.stopLoading();
```
Comment (collinjackson) [inline packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebView.java]: This looks like it might need to be removed.

File: packages/webview_flutter/example/test_driver/webview.dart [hunk 2]
```diff
@@ -54,6 +54,46 @@ void main() {
...
+      testWebViewError('https://httpstat.us/500', errorCompleter);
+      final WebViewError error = await errorCompleter.future;
+      expect(error, isNotNull);
+      expect(error.isConnectError, false);
+      expect(error.statusCode, 500);
+    });
+  });
+}
+
+Future<void> testWebViewError(
```
Comment (collinjackson) [inline packages/webview_flutter/example/test_driver/webview.dart]: From talking to Amir I get the sense that he prefers not to have helper methods in tests, and instead to have really verbose tests with a lot of code duplication.

File: packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebView.java (hunk 1 above)
Comment (hpoul) [reply packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebView.java]: 😳️good catch. removed it.

File: packages/webview_flutter/example/test_driver/webview.dart (hunk 2 above)
Comment (hpoul) [reply packages/webview_flutter/example/test_driver/webview.dart]: okay, i have inlined the helper method: https://github.com/flutter/plugins/pull/1389/commits/05444faa184d53c1a1d008488d17f7aa3498f5b2
if you figure it's easier to figure out whats going on here, it's fine with me.. my personal taste would be to have the helper method, because this seems so verbose that it's actually harder to grasp whats going on (although "whats going on" for me, is knowing where the differences are between inputs vs. outputs, with so much boiler plate it's actually hard to see the different cases being tested..)
if it's okay this way, okay.. otherwise i can also revert it and change it to `Future<WebViewError>` whatever fits better..

**Comment by hpoul** [comment]
@collinjackson thanks for the feedback 👍 i've bumped minor version number and added basic CHANGELOG and hopefully implemented your comments *waiting for tests*

**Comment by wwwdata** [comment]
Did you want to split this up into 2 pull requests? I guess the second one would then be for handling authorization requests? I am especially interested in that one from https://github.com/flutter/plugins/pull/1283

So my question is: Are you still working on this or not? If not I would need the callback for the authorization challenge anyways now for a project and could also help out to open a pull request and push it through.

**Comment by wwwdata** [comment]
@hpoul opened another PR for auth requests. Thanks a lot for your work, I just added some additional tests and could mostly re-use your stuff.

**Comment by hpoul** [comment]
closing this in favor of @wwwdata 's PR https://github.com/flutter/plugins/pull/2420
